source/non_matching/mesh_classifier.cc

```diff
@@ -0,0 +1,541 @@
+          fe_values.get_quadrature_points();
+
+        for (unsigned int i = 0; i < points.size(); i++)
+          local_levelset_values[i] = level_set->value(points[i]);
+      }
+
+
+
+      template <int dim>
+      void
+      AnalyticLevelSetDescription<dim>::get_local_level_set_values(
+        const typename Triangulation<dim>::active_cell_iterator &cell,
+        const unsigned int                                       face_index,
+        Vector<double> &local_levelset_values)
+      {
+        fe_face_values.reinit(cell, face_index);
+        const std::vector<Point<dim>> &points =
+          fe_face_values.get_quadrature_points();
+
+        for (unsigned int i = 0; i < points.size(); i++)
+          local_levelset_values[i] = level_set->value(points[i]);
+      }
+
+
+
+      template <int dim>
+      const hp::FECollection<dim> &
+      AnalyticLevelSetDescription<dim>::get_fe_collection() const
+      {
+        return fe_collection;
+      }
+
+
+
+      template <int dim>
+      unsigned int
+      AnalyticLevelSetDescription<dim>::active_fe_index(
+        const typename Triangulation<dim>::active_cell_iterator &) const
+      {
+        return 0;
+      }
+    } // namespace MeshClassifierImplementation
+  }   // namespace internal
+
+
+
+  using namespace internal::MeshClassifierImplementation;
+
+  template <int dim>
+  template <class VECTOR>
+  MeshClassifier<dim>::MeshClassifier(const DoFHandler<dim> &dof_handler,
+                                      const VECTOR &         level_set)
+    : triangulation(&dof_handler.get_triangulation())
+    , level_set_description(
+        std::make_unique<DiscreteLevelSetDescription<dim, VECTOR>>(dof_handler,
+                                                                   level_set))
+  {}
+
+
+
+  template <int dim>
+  MeshClassifier<dim>::MeshClassifier(const Triangulation<dim> &triangulation,
+                                      const Function<dim> &     level_set,
+                                      const FiniteElement<dim> &element)
+    : triangulation(&triangulation)
+    , level_set_description(
+        std::make_unique<AnalyticLevelSetDescription<dim>>(level_set, element))
+  {
+    // The level set function must be scalar.
+    AssertDimension(level_set.n_components, 1);
+    AssertDimension(element.n_components(), 1);
+  }
+
+
+
+  template <int dim>
+  void
+  MeshClassifier<dim>::reclassify()
+  {
+    initialize();
+
+    // Loop over all cells and determine category of all locally owned
+    // cells and faces.
+    for (const auto &cell : triangulation->active_cell_iterators())
+      {
+        if (cell->is_locally_owned())
+          {
+            for (unsigned int f = 0; f < GeometryInfo<dim>::faces_per_cell; ++f)
+              {
+                face_locations[cell->face(f)] =
+                  determine_face_location_to_levelset(cell, f);
+              }
+
+            cell_locations[cell] = determine_cell_location_to_levelset(cell);
```

Comment:
Would it be possible to determine the location of the cell from the location of the faces? At least for the case where at least one of the faces is intersected, we know that the cell is also intersected. We might only miss the case where the cut is completely inside the cell, so we might need that one as a fall-back.

Reply:
For most cells, all the faces will be either inside/outside, so this fall-back would happen quite often: `O(h^(-dim))`. If we we don't care about the case where the zero contour is entirely inside a cell, we could determine the cell location only from the face locations. Do you think that would be better?

Reply:
I think we should try to go with it. Maybe leave a comment regarding the possibility of the zero contour inside the cell (which can happen for higher order shape functions). For a Lagrange basis we could add the additional check that one node value is having another sign?
I or @bergbauer can then later look into the basis transformation with tensor products which might then again allow us to cheaply do the check in the Bernstein basis also inside the cell.